packages/data-context/src/sources/CloudDataSource.ts

```diff
@@ -29,117 +36,239 @@ const REMOTE_SCHEMA_URLS = {
   production: 'https://dashboard.cypress.io',
 }
 
-export interface CloudExecuteRemote extends RemoteExecutionRoot {
-  operationType: OperationTypeNode
-  query: string
-  document?: DocumentNode
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+type StartsWith<T, Prefix extends string> = T extends `${Prefix}${infer _U}` ? T : never
+type CloudQueryField = StartsWith<keyof NexusGen['fieldTypes']['Query'], 'cloud'>
+
+export interface CloudExecuteQuery {
+  document: DocumentNode
   variables: any
 }
 
+export interface CloudExecuteRemote extends CloudExecuteQuery {
+  operationType: OperationTypeNode
+  requestPolicy?: RequestPolicy
+  onUpdatedResult?: (data: any) => any
+}
+
+export interface CloudExecuteDelegateFieldParams<F extends CloudQueryField> {
+  field: F
+  args: core.ArgsValue<'Query', F>
+  ctx: DataContext
+  info: GraphQLResolveInfo
+}
+
+export interface CloudDataSourceParams {
+  fetch: typeof fetch
+  getUser(): AuthenticatedUserShape | null
+  logout(): void
+  onError(e: Error): void
+}
+
+/**
+ * The CloudDataSource manages the interaction with the remote GraphQL server
+ * It maintains a normalized cache of all data we have seen from the cloud and
+ * ensures the data is kept up-to-date as it changes
+ */
 export class CloudDataSource {
-  private _cloudUrqlClient: Client
+  #cloudUrqlClient: Client
 
-  constructor (private ctx: DataContext) {
-    this._cloudUrqlClient = this.reset()
+  constructor (private params: CloudDataSourceParams) {
+    this.#cloudUrqlClient = this.reset()
+  }
+
+  get #user () {
+    return this.params.getUser()
+  }
+
+  get #additionalHeaders () {
+    return {
+      'Authorization': this.#user ? `bearer ${this.#user.authToken}` : '',
+      'x-cypress-version': pkg.version,
+    }
   }
 
   reset () {
-    return this._cloudUrqlClient = createClient({
+    return this.#cloudUrqlClient = createClient({
       url: `${REMOTE_SCHEMA_URLS[cloudEnv]}/test-runner-graphql`,
       exchanges: [
         dedupExchange,
-        cacheExchange,
+        cacheExchange({
+          // @ts-ignore
+          schema: urqlSchema,
+          ...urqlCacheKeys,
+          resolvers: {},
+          updates: {
+            Mutation: {
+              _cloudCacheInvalidate: (parent, { args }: {args: Parameters<Cache['invalidate']>}, cache, info) => {
+                cache.invalidate(...args)
+              },
+            },
+          },
+        }),
+        errorExchange({
+          onError: (err, operation) => {
+            // Ignore the error when there's no internet connection or when the
+            // cloud session is not valid, we want to logout the user on the app
+            if (err.graphQLErrors.some((e) => e.message.includes('Unauthorized'))) {
```

Comment:
I think this should be checking `error.networkError?.message === 'Unauthorized'` as well